keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,733 @@
+host-network Pods to be endpoints of both IPv4 and IPv6 Services).
+
+Some methods in `utilnode` such as `GetNodeHostIP` and
+`GetPreferredNodeAddress` may be updated to have dual-stack versions
+that return multiple IPs. The existing versions are still useful on
+their own though.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TODO
+
+## Design Details
+
+### Implementation
+
+As before, when using an external cloud provider, the external
+provider deals with setting `Node.Status.Addresses`, and when using a
+legacy cloud provider or bare metal, kubelet deals with setting it.
+
+#### With a Cloud Provider (External or Legacy)
+
+When using an external cloud provider, kubelet will pass the value of
+`--node-ips` (explicit or defaulted) to the provider via a Node
+annotation. (If the provider finds that kubelet has not set the
+`--node-ips` annotation, then it should assume it's running against an
+old kubelet, and behave as it would have in previous releases.)
+
+Kubelet or the external cloud provider will now get the provisional
+list of IPv4 and IPv6 node addresses from the cloud provider-specific
+code (which will hopefully have been updated to return IPv6 node IPs)
+and then modify the list of node addresses as follows. ("Kubelet"
+below really means "Either kubelet or the external cloud provider".)
+
+  1. When using a legacy cloud provider, if the deprecated `--node-ip`
+     argument was passed, then kubelet will first do the filtering and
+     sorting associated with that.
+
+  2. If `--node-ips` has a single element, then all addresses of the
+     non-matching family will be removed from the list.
+
+  3. If there are no longer any `InternalIP` or `ExternalIP` elements
+     in the list, then the provided `--node-ips` is invalid, and
+     kubelet will error out.
+
+  4. If either element of `--node-ips` is a specific IP address, and
+     that IP address does not appear in the address list, then the
+     provided `--node-ips` is invalid and kubelet will error out.
+
+     Likewise, if the indicated IP could not possibly be chosen as the
+     Primary/Secondary Node IP because it is an `ExternalIP` and would
+     always be ignored in favor of an `InternalIP`, then the provided
+     `--node-ips` is invalid and kubelet will error out.
+
+  5. At this point kubelet will determine what the Primary Node IP and
+     (if relevant) Secondary Node IP would be based on the current
+     address list. If the results are compatible with what is
+     specified in `--node-ips`, then we are done.
+
+  6. Otherwise, kubelet will look for an element of the address list
```

Comment:
Does "address list" here refer to "network interfaces seen by the kubelet" or what's previously set in `Node.Status.Addresses` or something else?

Reply:
This set of steps here is basically talking about what will replace [lines 96-152 here](https://github.com/kubernetes/kubernetes/blob/v1.18.4/pkg/kubelet/nodestatus/setters.go#L96). "Address list" refers to "the provisional list of IPv4 and IPv6 node addresses from the cloud provider-specific code" that we started with above, which is the `cloudNodeAddresses` in that code.

Reply:
Actually it's not accurate to say that's replacing lines 96-152 since we're talking about both external and legacy providers. But anyway, it's talking about how we take what the cloud provider `NodeAddresses()` method returns, and turn it into what we set on `Node.Status.Addresses`.